src/transformers/cache_utils.py

```diff
@@ -192,246 +222,541 @@ def get_mask_sizes(self, cache_position: torch.Tensor, layer_idx: int) -> tuple[
         The masks are then prepared according to the given lengths (kv_length, kv_offset) and patterns (i.e. sliding_window, chunk_size),
         for each layer.
         """
-        query_length = cache_position.shape[0]
-        past_seen_tokens = self.get_seq_length()
-        kv_length = query_length + past_seen_tokens
-        return kv_length, 0
+        kv_length, kv_offset = self.layers[layer_idx].get_mask_sizes(
+            cache_position, self.key_cache[layer_idx], self.value_cache[layer_idx]
+        )
+        return kv_length, kv_offset
```

Comment:
Similarly here, once again to give you an idea of what I mean: the layers should be self-sufficient, and independent from each other. The Cache class should only do very minimal convenient wrapping around them, but not any data logic
```suggestion
        return self.layers[layer_idx].get_mask_sizes(cache_position)
```